drivers/pinctrl/pinctrl_npcx.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2022 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_pinctrl
+
+#include <assert.h>
+#include <drivers/gpio.h>
+#include <drivers/pinctrl.h>
+#include <soc.h>
+
+#include "soc_gpio.h"
+
+/* Driver config */
+struct npcx_pinctrl_config {
+	/* scfg device base address */
+	uintptr_t base_scfg;
+};
+
+/**
+ * Get group-gpios prop. elements from each child node of the pinctrl node which
+ * label is 'gpio_default' and set their configurations later.
+ */
+#define NPCX_IO_PAD_DEFAULT_LIST					\
+	{DT_FOREACH_CHILD_VARGS(DT_NODELABEL(gpio_default),		\
+				DT_FOREACH_PROP_ELEM, group_gpios,	\
+				Z_PINCTRL_NPCX_IO_PAD_CONF_INIT)}
+
+const pinctrl_soc_pin_t io_pad_conf[] = NPCX_IO_PAD_DEFAULT_LIST;
```

Comment:
The `gpio-default` node is not used by any of the example boards. You should add this support in a separate PR.  But we should discuss how this node relates to the Chromebook specific `named-gpios` node.

Reply:
I has added the ```gpio-default```  node in both ```npcx7m6fb_evb``` and ```npcx9m6f_evb``` example boards of this PR. 

Please refer to the board DT file [here](https://github.com/zephyrproject-rtos/zephyr/pull/43638/files#diff-52b6222516f8967ea9256d61408c141844549786916bd7c562e7e45a30de99a3R84).
There are four groups to collect GPIOs share the same pad configurations.

1. Group0: GPIOs to enable 1.8V detection
```
	group0 {
		/* Pads of i2c0_0 */
		group-gpios = <&gpiob 4 0>,
			      <&gpiob 5 0>;
		low-voltage-input;
	};
```
2. Group1: GPIOs to output high/low level
```
	group1 {
		group-gpios = <&gpio7 6 0>,
			      <&gpioc 5 0>;
		output-high; /* output-low */
	};
```
3. Group 2: GPIOs to configure input with PU/PD. We can also put unsed GPIOs here for better power consumption.
```
	group2 {
		group-gpios = <&gpioc 6 0>;
		bias-pull-up; /* bias-pull-down*/
	};
```

4. Group 3: GPIOs to output high/low with open-drain driving. Please notice the default driving type in npcx GPIO modules is push-pull. We needn't enable ```drive-push-pull``` property explicitly.
```
	group3 {
		group-gpios = <&gpioa 5 0>;
		drive-open-drain;
		output-high; /* output-low */
	};
```

Reply:
The example above initializes io-pad configurations before starting the kernel. But we still need the ```named-gpios``` node if the application configures the specific io-pad via gpio utilities during run-time.

Reply:
I would still prefer that that the `gpio-default` node and `gpio-groups` property gets added as a separate PR.  Adding support for setting default GPIO pin states is separate from the pinctrl support.

For chromebooks, we have a need to setup default GPIO states from the devicetree, but I would like to make sure the mechanism for doing this can be used on other SoCs.

Reply:
> I would still prefer that that the gpio-default node and gpio-groups property gets added as a separate PR. Adding support for setting default GPIO pin states is separate from the pinctrl support.

Will do it later.

> For chromebooks, we have a need to setup default GPIO states from the devicetree, but I would like to make sure the mechanism for doing this can be used on other SoCs.

The main purpose of introducing the ```gpio-default node``` and ```group-gpios``` property is to replace ```def-lvol-io-list``` node and unused pins in ```named-gpios``` node in the Chromebook projects. I admit it influences the board DT file a lot. Will discuss it in the next PR.

Reply:
Done. The implementation of ```gpio-default``` node and ```group-gpios``` has been move to the next [PR](https://github.com/MulinChao/zephyr/tree/npcx_psl_gpio).